vantage-agent/lib/charms/vantage_agent/v0/agent_snapper.py

```diff
@@ -0,0 +1,193 @@
+
+from ops import ConfigChangedEvent, InstallEvent, RemoveEvent, StartEvent, StopEvent, Object
+from ops.charm import CharmBase
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+
+logger = logging.getLogger()
+
+
+class AgentSnapper(Object):
+
+    def __init__(self, charm: CharmBase, snap_name: str):
+        super().__init__(charm, None)
+
+        self._charm = charm
+        self._snap_path = Path("/usr/bin/snap")
+        self._snap_name = snap_name
+
+        event_handler_bindings = {
+            self._charm.on.install: self.handle_install,
+            self._charm.on.config_changed: self.handle_config_changed,
+            self._charm.on.stop: self.handle_stop,
+            self._charm.on.start: self.handle_start,
+            self._charm.on.remove: self.handle_remove,
+        }
+        for event, handler in event_handler_bindings.items():
+            self._charm.framework.observe(event, handler)
+
+    ## Event Handlers
+    def handle_install(self, event: InstallEvent):
+        logger.debug(f"## Processing install event for {self._snap_name}.")
+        self._charm.unit.status = WaitingStatus(f"Installing snap for {self._snap_name}")
+        try:
+            self.install_snap()
+        except Exception as e:
+            logger.error(f"## Error installing {self._snap_name}: {e}")
+            self._charm.unit.status = BlockedStatus("Error installing the snap for {self._snap_name}")
+            event.defer()
+            return
+
+        logger.debug(f"Snap for {self._snap_name} installed")
+
+    def handle_config_changed(self, event: ConfigChangedEvent):
+        logger.debug(f"## Processing config changed event for {self._snap_name}.")
+        if not self.is_snap_installed():
+            logger.debug(f"## Snap for {self._snap_name} not installed, deferring event")
+            event.defer()
+            return
+
+        logger.info(f"## Updating config for {self._snap_name}")
+        self.configure_snap()
+
+        if self.is_snap_active():
+            self.run_snap_service("restart")
+
+    def handle_start(self, event: StopEvent):
+        logger.debug(f"## Processing start event for {self._snap_name}.")
+        self.run_snap_service("start")
+        self._charm.unit.status = ActiveStatus()
+
+    def handle_stop(self, event: StopEvent):
+        logger.debug(f"## Processing stop event for {self._snap_name}.")
+        self.run_snap_service("stop")
+        self._charm.unit.status = MaintenanceStatus()
+
+    def handle_remove(self, event: RemoveEvent):
+        logger.debug(f"## Processing remove event for {self._snap_name}.")
+        self.remove_snap()
+
+    ## Operations
+    def is_snap_installed(self) -> bool:
+        cmd = [
+            self._snap_path,
+            "list",
+            self._snap_name,
+        ]
+
+        try:
+            subprocess.check_output(cmd)
+            return True
+        except subprocess.CalledProcessError:
+            return False
```

Comment:
not that it make a big difference, but it seems to me the case for using a `property` decorator. But anyway, the behaviour would be the same throughout the code

Reply:
Resolved in commit e15fece.